File: Prob1.py
# Method 1 - Hashmap
# Time Complexity - O(n), where n is the length of s
# Space Complexity - O(n) 
# We use a running sum to emulate the generation of all possible contiguous arrays.
# If running sum is in the hashmap, increase count else make a new entry with 1 as value .
# If complement present in running sum, add value at dict[complement] to count.
# key adjustment for calculations - dict[0]=1

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count=0
        running_sum=0
        running_sum_dict=defaultdict(lambda:0)
        running_sum_dict[0]=1           #key adjustment for calculations
        for i in range(len(nums)):
            running_sum+=nums[i]
            complement=running_sum-k
            if complement in running_sum_dict:
                count+=running_sum_dict[complement]
            running_sum_dict[running_sum]+=1
        return count

File: prob1.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes


class Solution {
    public int subarraySum(int[] nums, int k) {
        int sum = 0;
        int ans = 0;
        HashMap<Integer,Integer> map = new HashMap<>();
        map.put(0,1);
        for(int j=0;j<nums.length;j++){
            sum += nums[j];
            if(map.containsKey(sum -k)){
                ans += map.get(sum-k);
            }
            map.put(sum,map.getOrDefault(sum,0)+1);
        }
        return ans;
    }
}